File: Dev/C++/Mascaret/src/MascaretApplication.cpp
#include "Tools/filesystem.h"
#ifdef _WIN32
#include "winsock2.h" // for gethostname
#endif
#include <stdio.h>
//#include <netdb.h>
#include <iostream>
#include <fstream>

//#include <pthread.h>

#include "MascaretApplication.h"

#include "Tools/Loaders/InstanceLoader.h"
#include "Tools/Loaders/ModelLoader.h"
#include "Tools/Loaders/ModelLoader2.h"
#include "VEHA/Behavior/Common/pluginManager.h"
#include "BEHAVE/AgentClass.h"
#include "BEHAVE/Agent.h"
#include "BEHAVE/AgentPlateform.h"
#include "VEHA/Entity/VirtualRealityComponentFactory.h"
#include "VEHA/Kernel/PrimitiveType.h"
#include "IEHA/Visualisation/Camera.h"
#include "IEHA/Visualisation/Window3D.h"
//#include "MATS/Poseidon/LearningSession.h"

#include "VEHA/Behavior/Common/behaviorScheduler.h"
#include "HAVE/Body.h"
#include "VEHA/Entity/EntityClass.h"

#include "Tools/Scenario/ScenarioScheduler.h"
#include "Tools/Scenario/Scenario.h"

DECLARE_NAMESPACE_FILESYSTEM;

typedef void (*fct)(void);

MascaretApplication* MascaretApplication::_instance = NULL;

MascaretApplication::MascaretApplication()
{
  cerr << "Mascaret build datetime: " << __DATE__ << " " << __TIME__ << endl;
  if (_instance) throw runtime_error("Une autre instance de MascaretApplication a déjà été créée");
  _instance = this;
  char hostname[1024];
  gethostname(hostname, 1024);
  _hostname = string(hostname);
//	_learningSession = shared_ptr<Poseidon::LearningSession>(new Poseidon::LearningSession());
}

MascaretApplication::~MascaretApplication()
{
  PluginManager::getInstance()->fireMascaretDestroying();
  BehaviorScheduler::destroyInstance();
}

void MascaretApplication::parse(const string& applicationFileName)
{
  cerr << "[MascaretApplication Info] #### Start parsing application file: " << applicationFileName << endl;
  fs::path appPath(fs::complete(applicationFileName));
#if BOOST_FILESYSTEM_VERSION == 2
  _baseDir=appPath.branch_path().file_string();
  cerr << "  name: "<< appPath.branch_path().file_string() << endl;
#else
  _baseDir = appPath.branch_path().string();
  cerr << "  name: " << appPath.branch_path().string() << endl;
#endif

  cerr << "[MascaretApplication Info] Creating parser object...";
  shared_ptr<XmlParser> parser = make_shared<XmlParser>();
  cerr << " OK!" << endl;
  parser->parseFile(applicationFileName);
  cerr << "[MascaretApplication Info] #### Finish parsing application file: " << applicationFileName << " OK!" << endl;
  shared_ptr<XmlNode> root = parser->getRoot();

  parse(root);

  // Init Plugin
  PluginManager::getInstance()->fireMascaretModelLoaded();
}
void MascaretApplication::parse(shared_ptr<XmlNode> root)
{
  shared_ptr<XmlNode> appliNode = root->getChildByName("ApplicationParameters");
  shared_ptr<XmlNode> actNode = root->getChildByName("Actors");
  shared_ptr<XmlNode> orgNode = root->getChildByName("Organisations");

  if (appliNode)
  {
    parseApplication(appliNode);
  }
  else
  {
    // Parametres par defaut de l'applie ....
    string ressourceDir = "HTTPServerBaseDir";
    agentPlateform = boost::make_shared<AgentPlateform>(_hostname, 8080, ressourceDir);
    agent = boost::make_shared<Agent>(agentPlateform, "HTTPManager", shared_ptr<AgentClass>());
    agent->addBehavior("SimpleCommunicationBehavior");
    agentPlateform->addAgent(agent);
    ////createServer(new HTTPManager(ressourceDir,agentPlateform,agent),8080);
  }

  //parseEnvironment(model,root);

  shared_ptr<XmlNode> envChild = root->getFirstChild();
  while (envChild)
  {
    if (envChild->getName() == "Environment")
    {
      string envUrl = envChild->getProperty("url");
      shared_ptr<Environment> env = parseEnvironment(envUrl, actNode, orgNode);
    }
    else if(envChild->getName() == "Scenario")
    {
        string url = envChild->getProperty("url");
#if BOOST_FILESYSTEM_VERSION == 2
        string fullpath = fs::complete(url,_baseDir).file_string();
#else
        string fullpath = fs::complete(url, _baseDir).string();
#endif
        Scenario::ScenarioScheduler::getInstance()->parseScenarioFile(fullpath);
        string scenario = Scenario::ScenarioScheduler::getInstance()->getScenario()->writeScenario();
        cout << scenario << endl;
    }
    envChild = envChild->getNext();
  }
}
void MascaretApplication::init()
{
  //Nothing to do
}
void MascaretApplication::run()
{
  while (step())
  {
  }
}
bool MascaretApplication::step()
{
  BehaviorScheduler::getInstance()->step();
  Scenario::ScenarioScheduler::getInstance()->activate();
  PluginManager::getInstance()->fireMascaretStep();
  return true;
}
void MascaretApplication::generateDocumentation()
{
//TODO
}

void MascaretApplication::parseApplication(shared_ptr<XmlNode> appliNode)
{
  bool noServer = false;
  string ressourceDir = "../HTTPServerBaseDir";
  int port = 8080;
  int ports = 8080;
  char _hostname[1024];

  gethostname(_hostname, 1024);

  shared_ptr<XmlNode> serverNode = appliNode->getChildByName("HTTP");
  if (serverNode)
  {
    if (serverNode->hasProperty("NoServer")) serverNode->getPropertyBoolean("httpPort", noServer);
    if (serverNode->hasProperty("RessourceDir")) ressourceDir = serverNode->getProperty("RessourceDir");
    if (serverNode->hasProperty("httpPort")) serverNode->getPropertyInteger("httpPort", port);
    if (serverNode->hasProperty("httpsPort")) serverNode->getPropertyInteger("httpsPort", ports);
  }

  agentPlateform = boost::make_shared<AgentPlateform>(_hostname, port, ressourceDir, noServer);
  shared_ptr<XmlNode> userNode = appliNode->getChildByName("User");
  if (userNode)
  {
    agent = boost::make_shared<Agent>(agentPlateform, userNode->getProperty("name"), shared_ptr<AgentClass>());
    agentPlateform->addAgent(agent);
    agent->addBehavior("SimpleCommunicationBehavior");
  }
  else
  {
    agent = make_shared<Agent>(agentPlateform, "HTTPManager", shared_ptr<AgentClass>());
    agent->addBehavior("SimpleCommunicationBehavior");
    agentPlateform->addAgent(agent);
  }
  //createServer(new HTTPManager(ressourceDir,agentPlateform, agent),port);

  shared_ptr<XmlNode> plugsNode = appliNode->getChildByName("Plugins");
  if (plugsNode)
  {
    cerr << "Loading plugins..." << endl;
#if BOOST_FILESYSTEM_VERSION == 2
    PluginManager::getInstance()->setPluginDir(fs::complete(plugsNode->getProperty("PluginsDir"),_baseDir).file_string());
#else
    PluginManager::getInstance()->setPluginDir(
        fs::complete(plugsNode->getProperty("PluginsDir"), _baseDir).string());
#endif

    //PluginManager::getInstance()->setPluginDir(plugsNode->getProperty("PluginsDir"));
    shared_ptr<XmlNode> pluginNode = plugsNode->getFirstChild();
    while (pluginNode)
    {
      if (pluginNode->getName() == "Plugin")
      {
        string name = pluginNode->hasProperty("Name") ? pluginNode->getProperty("Name") : pluginNode->getProperty("name");
        PluginManager::getInstance()->openPlugin(name, pluginNode);
      }
      pluginNode = pluginNode->getNext();
    }
  }

}

shared_ptr<Model> MascaretApplication::parseModel(string url)
{
#if BOOST_FILESYSTEM_VERSION == 2
  shared_ptr<ModelLoader2> modelLoader =boost::make_shared<ModelLoader2>(fs::complete(url,_baseDir).file_string());
#else
  shared_ptr<ModelLoader2> modelLoader = make_shared<ModelLoader2>(fs::complete(url, _baseDir).string());
#endif
  shared_ptr<Model> model = modelLoader->getModel();
  model->setModelLoader(modelLoader);

  return model;
}

shared_ptr<Environment> MascaretApplication::parseEnvironment(string url,
                                                              shared_ptr<XmlNode> actNode,
                                                              shared_ptr<XmlNode> orgNode)
{
  shared_ptr<Environment> env;

  shared_ptr<XmlParser> parser = make_shared<XmlParser>();
  parser->parseFile(url);
  shared_ptr<XmlNode> root = parser->getRoot();

  shared_ptr<Model> model;
  shared_ptr<XmlNode> modelNode = root->getChildByName("Model");
  if (modelNode)
  {
    string urlModel = modelNode->getProperty("url");

    model = parseModel(urlModel);

    map<string, shared_ptr<Environment> >::iterator it = model->getEnvironments().find(url);
    if (model->getEnvironments().size() != 0 && it != model->getEnvironments().end())
    {
      return it->second;
    }
    else
    {
      env = make_shared<Environment>(model);
      env->setUrl(url);
      model->addEnvironment(env);

#if BOOST_FILESYSTEM_VERSION == 2
      InstanceLoader instanceLoader(agentPlateform, env, fs::complete(url,_baseDir).file_string());
#else
      InstanceLoader instanceLoader(agentPlateform, env, fs::complete(url, _baseDir).string());
#endif

      cout << "SIZE of env map " << model->getEnvironments().size() << " of model :"
          << model->getUrl() << endl;

      if (actNode)
#if BOOST_FILESYSTEM_VERSION == 2
        instanceLoader.parseInstances(agentPlateform,env,fs::complete(actNode->getProperty("url"),_baseDir).file_string());
#else
        instanceLoader.parseInstances(agentPlateform, env,
                                      fs::complete(actNode->getProperty("url"), _baseDir).string());
#endif

      if (orgNode)
#if BOOST_FILESYSTEM_VERSION == 2
        instanceLoader.parseInstances(agentPlateform,env,fs::complete(orgNode->getProperty("url"),_baseDir).file_string());
#else
        instanceLoader.parseInstances(agentPlateform, env,
                                      fs::complete(orgNode->getProperty("url"), _baseDir).string());
#endif
    }
  }
  else
    cerr << "There's no model in this environment.Please add a model in " << url << "." << endl;
  return env;
}

shared_ptr<PrimitiveType> MascaretApplication::getBasicType(string type)
{

  return getModel()->getBasicType(type);
  /*

   shared_ptr<PrimitiveType> primitiveType;

   map<string,shared_ptr<Model> >::iterator it;
   for(it = _models.begin();it!=_models.end();it++)
   {else
   cerr<<"There's no model in this environment.Please add a model in "<<url<<"."<<endl;
   primitiveType = it->second->getBasicType(type);
   }

   return primitiveType;*/

}

shared_ptr<Environment> MascaretApplication::getEnvironment()
{
  return _models.begin()->second->getEnvironments().begin()->second;
}

shared_ptr<Model> MascaretApplication::getModel()
{
  return _models.begin()->second;
}

shared_ptr<Model> MascaretApplication::getModel(string modelName)
{
  map<std::string, boost::shared_ptr<Model> >::iterator it = _models.find(modelName);
  if (it != _models.end())
    return it->second;
  else
    return shared_ptr<Model>();
}

bool MascaretApplication::hasModel(string modelName)
{
  map<std::string, boost::shared_ptr<Model> >::iterator it = _models.find(modelName);
  return it != _models.end();
}

File: Dev/CS/Mascaret/Mascaret/MascaretApplication.cs

using System;
using System.IO;
using System.Xml.Linq;
using System.Collections;
using System.Collections.Generic;


namespace Mascaret
{

    public delegate void LogHandler(string message);

    public class MascaretApplication
    {
        protected static MascaretApplication instance = null;

        public static MascaretApplication Instance
        {
            get
            {
                if (instance == null)
                {
                    instance = new MascaretApplication();
                }
                return instance;
            }
        }

        #region Log
        public event LogHandler Log;

        public void OnLog(string message)
        {
            if (Log != null)
                Log(message);
        } 
        #endregion

        public virtual void parse(string applicationFileName, string baseDir)
        {
           
          
            BaseDir =baseDir;

            XDocument parser;
           // logfile.WriteLine("Loading application file : " + BaseDir + "/" + applicationFileName); logfile.Flush();
            string assetPath = BaseDir + "/"+applicationFileName;


            String s = readFlow(BaseDir + "/" + applicationFileName);
            parser = XDocument.Parse(s);
            XElement root = parser.Root;
            parse(root);
        }

        public virtual void parse(XElement root)
        {
            XElement appliNode = root.Element("ApplicationParameters");
            XElement actnode = root.Element("Actors");
            XElement orgNode = root.Element("Organisations");

            {
                if (agentPlateform == null)
                {
                    this.VRComponentFactory.Log("Agent Plateform");
                    // Parametres par defaut de l'applie ....
                    string ressourceDir = "HTTPServerBaseDir";
                    agentPlateform = new AgentPlateform("localhost", 8080, ressourceDir, false);
                    agent = new Agent(agentPlateform, "HTTPManager", null, "");
                    agent.addBehavior("SimpleCommunicationBehavior", 0.2, true);
                    agentPlateform.addAgent(agent);
                }
     
            }

            foreach (XElement child in root.Elements())
            {

                if (child.Name.LocalName == "Environment")
                {
                    string envUrl = child.Attribute("url").Value;
                    Environment env = parseEnvironment(envUrl, actnode, orgNode);
                }
            }
        }

        Model parseModel(string url)
        {
            //StreamWriter file = new StreamWriter("log.txt");
           // file.WriteLine(url);
           // file.Flush();
           // logfile.WriteLine("Parsing model file : " + url); logfile.Flush();
            ModelLoader2 modelLoader = new ModelLoader2(url, true);
            Model model = modelLoader.Model;
            model.ModelLoader = modelLoader;
            //file.Close();

            return model;
        }

        Environment parseEnvironment(string url, XElement actNode, XElement orgNode)
        {
            Environment env = null;

         //   logfile.WriteLine("Parsing Environment file : " + url); logfile.Flush();

            String s = readFlow(BaseDir + "/" + url);
            XDocument parser = XDocument.Parse(s);
            XElement root = parser.Root;

        //    Model model;
            XElement modelNode = root.Element("Model");
            if (modelNode != null)
            {
                string urlModel = modelNode.Attribute("url").Value;

               /// StreamWriter file = new StreamWriter("log2.txt");
               // file.WriteLine(" URL : " + url + " : " + BaseDir + "/" + urlModel);
              //  file.Flush();
              //  file.Close();
                if (model == null) model = parseModel(BaseDir+"/"+urlModel);
                
               // if (model.Environments.ContainsKey(url))
              //  {
               //     env = model.Environments[url];
              //  }
              //  else
                {
                    env = new Environment(model);
                    env.Url = url;
                    model.addEnvironment(env);

                    InstanceLoader instanceLoader = new InstanceLoader(agentPlateform, env, BaseDir+"/"+url, true);

                    System.Console.WriteLine("SIZE of env map " + model.Environments.Count + " of model :" + model.name);

                    if (actNode != null)
                        instanceLoader.parseInstances(agentPlateform, env, actNode.Attribute("url").Value, true);

                    if (orgNode != null)
                        instanceLoader.parseInstances(agentPlateform, env, orgNode.Attribute("url").Value, true);
                }
                
            }
            else
                env = null;
            return env;
        }




        private Model model = null;
        public Model Model
        {
            get { return model; }
            set { model = value; }
        }

        protected Dictionary<string, Model> models;
        public Dictionary<string, Model> Models
        {
            get { return models; }
            set { models = value; }
        }

        private AgentPlateform agentPlateform = null;
        public AgentPlateform AgentPlateform
        {
            get { return agentPlateform; }
            set { agentPlateform = value; }
        }

        private Agent agent = null;
        public Agent Agent
        {
            get { return agent; }
            set { agent = value; }
        }

        public Environment getEnvironment()
        {
            if (Model.Environments != null && Model.Environments.Keys.Count > 0)
            {
                IEnumerator<KeyValuePair<string, Environment>> iter = Model.Environments.GetEnumerator();
                iter.MoveNext();
                return (Environment)(iter.Current.Value);
            }
            else
            {
                Environment env = new Environment(Model);
                env.Url = "default";
                Model.addEnvironment(env);
                return env;
            }
        }

        protected string baseDir = "";
        public string BaseDir
        {
            get { return baseDir; }
            set { baseDir = value; }
        }

        public bool hasModel(string modelName)
        {
            if (model != null && model.name == modelName)
                return true;

            if (models != null)
            {
                foreach (Model currentModel in models.Values)
                {
                    if (currentModel.name == modelName)
                        return true;
                }
            }
            return false;
        }

        public Model getModel(string modelName)
        {
            if (model != null && model.name == modelName)
                return model;

            if (models != null)
            {
                foreach (Model currentModel in models.Values)
                {
                    if (currentModel.name == modelName)
                        return currentModel;
                }
            }
            return null;
        }

        public void addModel(Model model)
        {
            if (models == null)
            {
                models = new Dictionary<string, Model>();
            }

            models.Add(model.name, model);
            if (this.model == null)
                this.model = model;
        }

        public MascaretApplication()
        {
            instance = this;
        }

        public void init()
        {
           
        }

        public void run()
        {
            while (step())
            {
            }
        }

        public bool step()
        {
            BehaviorScheduler.Instance.Update();
            //this.httpCommunicationManager.httpServer.Update();

            return true;
        }

        private VirtualRealityComponentFactory virtualRealityComponentFactory;
        public VirtualRealityComponentFactory VRComponentFactory
        {
            get { return virtualRealityComponentFactory; }
            set { virtualRealityComponentFactory = value; }
        }

        public string readFlow(string url)
        {
            return virtualRealityComponentFactory.readFlow(url);
        }


    }
}
